runtime/compiler/compile/J9SymbolReferenceTable.cpp

```diff
@@ -608,47 +617,206 @@ J9::SymbolReferenceTable::findOrFabricateShadowSymbol(TR::ResolvedMethodSymbol *
+      qualifiedFieldName,
+      TR::Symbol::UnknownField);
+
+   mcount_t methodIndex = mcount_t::valueOf(0);
+   int32_t cpIndex = -1;
+   symRef = new (trHeapMemory()) TR::SymbolReference(
+      self(),
+      sym,
+      methodIndex,
+      cpIndex);
+
+   bool isResolved = true;
+   bool isUnresolvedInCP = false;
+   initShadowSymbol(NULL, symRef, isResolved, type, offset, isUnresolvedInCP);
+
+   _resolvedFieldShadows.insert(std::make_pair(key, symRef));
+   return symRef;
+   }
+
+TR::Symbol *
+J9::SymbolReferenceTable::createShadowSymbol(
+   TR::DataType type,
+   bool isVolatile,
+   bool isPrivate,
+   bool isFinal,
+   const char *name,
+   TR::Symbol::RecognizedField recognizedField)
+   {
+   TR::Symbol *sym = NULL;
+   if (recognizedField != TR::Symbol::UnknownField)
+      sym = TR::Symbol::createRecognizedShadow(trHeapMemory(), type, recognizedField);
+   else
+      sym = TR::Symbol::createShadow(trHeapMemory(), type);
+
+   if (name != NULL)
+      {
+      sym->setNamedShadowSymbol();
+      sym->setName(name);
+      }
+
+   if (isVolatile)
+      sym->setVolatile();
+
+   if (isPrivate)
+      sym->setPrivate();
+
+   if (isFinal)
+      sym->setFinal();
+
+   static char *dontAliasShadowsToEarlierGISEnv = feGetEnv("TR_dontAliasShadowsToEarlierGIS");
+   bool dontAliasShadowsToEarlierGIS = dontAliasShadowsToEarlierGISEnv != NULL;
+
+   if (aliasBuilder.mutableGenericIntShadowHasBeenCreated() && !dontAliasShadowsToEarlierGIS)
+      {
+      // Some previously-created GIS might actually refer to this shadow
+      aliasBuilder.setConservativeGenericIntShadowAliasing(true);
+      }
+
+   return sym;
+   }
 
 TR::SymbolReference *
 J9::SymbolReferenceTable::findOrCreateShadowSymbol(TR::ResolvedMethodSymbol * owningMethodSymbol, int32_t cpIndex, bool isStore)
    {
-   TR_ResolvedMethod * owningMethod = owningMethodSymbol->getResolvedMethod();
+   TR_ResolvedJ9Method * owningMethod =
+      static_cast<TR_ResolvedJ9Method*>(owningMethodSymbol->getResolvedMethod());
+
    bool isVolatile = true, isFinal = false, isPrivate = false, isUnresolvedInCP;
    TR::DataType type = TR::NoType;
    uint32_t offset = 0;
    bool resolved = owningMethod->fieldAttributes(comp(), cpIndex, &offset, &type, &isVolatile, &isFinal, &isPrivate, isStore, &isUnresolvedInCP, true);
    bool sharesSymbol = false;
 
+   TR_OpaqueClassBlock *containingClass = NULL;
+   TR::Symbol::RecognizedField recognizedField = TR::Symbol::searchRecognizedField(comp(), owningMethod, cpIndex, false);
+
+   if (isStore && isPrivate && !comp()->getOptions()->realTimeGC() &&
+       owningMethodSymbol->getResolvedMethod()->getRecognizedMethod() == TR::java_lang_ref_SoftReference_get &&
+       recognizedField == TR::Symbol::Java_lang_ref_SoftReference_age)
+      {
+      isVolatile = false;
+      }
+
+   if (resolved)
+      {
+      bool isStatic = false;
+      containingClass =
+         owningMethod->definingClassFromCPFieldRef(comp(), cpIndex, isStatic);
```

Comment:
I was about to post a comment about how you're missing a validation record here. However, the call above to `TR_ResolvedRelocatableJ9Method::fieldAttributes` calls `TR_ResolvedJ9Method::definingClassFromCPFieldRef` and then proceeds to add a validation record.

That said, this is very fragile, and probably dangerous. Therefore, you should explicitly add the validation record in `TR_ResolvedRelocatableJ9Method::definingClassFromCPFieldRef`, However, because the query seems to be the same as what's done in `fieldAttributes`, I believe you can reuse the existing validation:
```
TR_OpaqueClassBlock *
TR_ResolvedRelocatableJ9Method::definingClassFromCPFieldRef(
   TR::Compilation *comp,
   I_32 cpIndex,
   bool isStatic)
   {
   TR_OpaqueClassBlock *clazz = TR_ResolvedJ9Method::definingClassFromCPFieldRef(comp, cp(), cpIndex, isStatic);

   bool valid = false;
   if (comp->getOption(TR_UseSymbolValidationManager))
      valid = comp->getSymbolValidationManager()->addDefiningClassFromCPRecord(clazz , cp(), cpIndex, isStatic);
   else
      valid = storeValidationRecordIfNecessary(comp, cp(), cpIndex, isStatic ? TR_ValidateStaticField : TR_ValidateInstanceField, persistentIdentifier());

   if (!valid)
      clazz = NULL;

   return clazz;
   }
```
You can keep the `TR_ASSERT_FATAL` as is below since in order to have gotten to this point, the call to `fieldAttributes` above must have succeeded.

Reply:
Fixed in cece2b9.